### ISSUE TITLE
Validation of GNILC dust templates

### ISSUE THREAD

**Issue by zonca** [galsci/pysm#101]
The implementation of the GNILC-based dust templates is complete, see details and the executed notebook at:

https://github.com/galsci/pysm/pull/97#issuecomment-1006313372

Now it would be useful to have some independent validation, load the maps, take spectra, compare known regions, check everything is reasonable (see also #100)

**Comment by zonca** [comment]
## Files location

https://portal.nersc.gov/project/cmb/pysm-dev/gnilc/

or at NERSC in:

/global/project/projectdirs/cmb/www/pysm-dev/gnilc

**Comment by zonca** [comment]
# Available files

## Output templates

    gnilc_dust_template_nside2048_float32.fits
    gnilc_dust_template_nside4096_float32.fits
    gnilc_dust_template_nside8192_float32.fits

## Input GNILC products

For temperature:

    COM_CompMap_Dust-GNILC-F353_2048_21p8acm.fits

For polarization:

    COM_CompMap_IQU-thermaldust-gnilc-varres_2048_R3.00.fits

## Masks

    BK15_region_Gal_apo.fits
    HFI_Mask_GalPlane-apo2_2048_R2.00.fits

## Intermediate products

Large scale Alms in logpoltens

    gnilc_dust_largescale_template_logpoltens_alm_nside2048_lmax3072_complex64.fits

Small scales power spectrum

    gnilc_dust_small_scales_logpoltens_cl_lmax16384.fits

Modulation of the small scales

    gnilc_dust_temperature_modulation_alms_lmax3072.fits
    gnilc_dust_polarization_modulation_alms_lmax3072.fits

**Comment by zonca** [comment]
@brandonshensley @seclark @NicolettaK @yaojian95 @giuspugl @delabru

**Comment by zonca** [comment]
## Spectral index and Dust temperature

At the same location above I have also copied the outputs of the GNILC beta and Td processing, see the notebooks in https://github.com/galsci/pysm/pull/104 for more details.

### Output maps

6 maps, 2 componets at 3 resolutions:

`gnilc_dust_[beta,Td]_nside[2048,4096,8192]_float32.fits`

### Intermediate products

```
gnilc_dust_largescale_template_beta_alm_nside2048_lmax1024_complex64.fits
gnilc_dust_largescale_template_Td_alm_nside2048_lmax1024_complex64.fits
gnilc_dust_small_scales_beta_cl_lmax16384.fits
gnilc_dust_small_scales_Td_cl_lmax16384.fits
```

For modulation I reuse the "temperature" modulation above.

**Comment by bthorne93** [comment]
A while ago I was calculating power spectra of the new log pol tens templates. I found that the power spectra of smaller patches of sky exhibits more of a break at the transition scale around ell of 100 on smaller patches of sky. For example,  the bottom panel of this plot:

![Screen Shot 2022-03-17 at 11 52 59 AM](https://user-images.githubusercontent.com/16899444/158875299-a0f055fa-473e-4355-a1f6-578eef12fb3e.png)

The bottom panel shows BB, and the green line shows the spectra calculated on the BICEP / Keck patch, with the dashed line being the input map `COM_CompMap_Dust-GNILC-F353_2048_21p8acm.fits`, and the solid line the derived log pol tens template, at the time. Initially I was concerned about this plot because we would be under predicting dust BB in the BK patch in the important ell range of 100-300 by almost an order of magnitude.   

Andrea and Giuseppe were unable to reproduce these plots, and so I reran the spectra, using the updated log pol tens templates in this thread, as well as the BK mask in this thread: `BK15_region_Gal_apo.fits`. This is a comparison of the various masks (Andrea's BK mask is labelled `BK_NERSC`): 

![validation_masks (1)](https://user-images.githubusercontent.com/16899444/158875758-c68a6605-3143-4d53-b6ee-e6baaab59f62.png)

With the updated templates, I find that the discrepancy in the BK patch (both my original mask, and the new mask), is reduced by a factor of a few, whilst the BB spectrum in the LR15 region stays a little low (and strangely lower than the two BK regions). I've also plotted in the BB panel the maximum likelihood BK dust model for comparison. 

![spectra_by_field_ana_NSIDE1024 (7)](https://user-images.githubusercontent.com/16899444/158871943-76166ced-9159-464c-be3f-fdff6a57d4db.png)

Based on this final plot, I'm not sure this is an issue worth investigating much further, but I wanted to record it here.

**Comment by giuspugl** [comment]
Thanks a lot @bthorne93  for this follow up! it is great to see this validation with the latest updated models! 

1. Firstly a  comment on what has changed between the two versions of plots is that in the former one we had injected small scales using the  fitted  power law  indices  from TT, EE and BB spectra (respectively about  -1.2,    -0.3,-0.4 ). In the latter,  the small scales are included adopting  the same TT index for all the IQU maps. This results in bumping up EE and BB modes in the final spectra (so the _edge effect_ is milder than the former plot) , but it is not intuitive to me how this translate into effectively having less power in LR15 than BK patches . Any clue? 
2. about the _edge _ we observe especially in EE and BB from LR34 and below i think that we can't do much more than what we currently have done.. It 's great to see that from the  spectra  going down very well  with a power law (remember we modulate the small scales so that we have something reliable at fsky=80%) up to fsky=20% is a great achievement, but was expecting the _edge_ to be more visible for smaller patches . This points out that we really need to implement a modulation methodology that is also local (somehow with needlets or something similar) .
3. lastly thanks a lot for including  the solid black line from BK max. likelihood. It adds an extra-data point in our understanding:  the BK estimates of dust were  essentially driven by the dust spectrum from GNILC maps at ell<100, the spectral index seems to me kinda parallel to the dashed green and orange lines, but the BK estimated amplitude seems to me factor of 2 lower than the one from the GNILC template. Do you know why ?

**Comment by bthorne93** [comment]
@giuspugl 

1. I am not sure how fixing the power law index to the TT value, which is steeper than the other two values you showed, would bump up the EE/BB spectra at multipoles higher than the pivot scale (which I think is ~100, right?). I would have thought that change would actually make the simulated power lower at scales above the pivot scale. 
3. Since we are fixing the slope of the power spectrum to the measured TT value of -1.2, I think it makes sense that we see a much steeper slope in BB than BICEP / Keck measured. The BICEP / Keck dataset has higher S/N on the dust amplitude in their 220 GHz channel, rather than in the Planck 353 GHz data. Perhaps in this region we have low S/N, resulting in slightly biased autospectra. This may also explain why there is very little difference between the BK and LR15 regions?

**Comment by brandonshensley** [comment]
@giuspugl The issue of alpha_BB is being raised, so I think it may be worth digging back into @bthorne93's question 1 above.

**Comment by giuspugl** [comment]
>I am not sure how fixing the power law index to the TT value, which is steeper than the other two values you showed, would bump up the EE/BB spectra at multipoles higher than the pivot scale (which I think is ~100, right?). I would have thought that change would actually make the simulated power lower at scales above the pivot scale.

@bthorne93 , sorry I totally lost track of this comment! I totally agree that choosing a steeper spectr. index would lower the power than a flat one. (i think there was a typo in my previous comment). 
 A solution we can discuss is to set 2 pivot scales (`ell_p1 ~100` and `ell_p2 ~600` )  : 
 
- if `ell<ell_p1` : we  use GNILC  templates,
- if `ell_p1<ell <ell_p2` use flat  spectral index consistent w/ BK  and Planck data
- if `ell> ell_p2` use a steep  spectral index to avoid crossing of EE and BB over TT 
  

This approach is still in line with our philosophy, i.e. use as much as we could observations  and extend simulations with physical assumptions.

**Comment by seclark** [comment]
I like this approach. The steeper spectral index should be chosen to anticipate high nside. Maybe we select the index to ensure that EE, BB < TT out to 8192 at least.

**Comment by giuspugl** [comment]
In this [notebook](https://gist.github.com/giuspugl/4a6eb0d600df97c1bb5c44451f0c542e) you'll find an attempt to implement what i proposed above: 

> * if `ell<ell_p1` : we  use GNILC  templates,
> * if `ell_p1<ell <ell_p2` use flat  spectral index consistent w/ BK  and Planck data
> * if `ell> ell_p2` use a steep  spectral index to avoid crossing of EE and BB over TT

Even in this case, our benchmark is to match small and large scales for the mask `GAL080`, however this 

/!\ This new implementation employs several new features wrt the previous `d9, d10`  models: 
- two pivot scales `ell_p1,ell_p2`  for small scale injection 
-  spectral indices for  EE, BB coming from literature ,   [Planck 2018 XI](https://arxiv.org/pdf/1801.04945.pdf) and for TT from  [Miville-Deschenes 2016](https://www.aanda.org/articles/aa/pdf/2016/09/aa28503-16.pdf)
- we inject small scales with non-zero TE from [Planck 2018 XI](https://arxiv.org/pdf/1801.04945.pdf)
- Inside the `GAL097` mask (i.e. along the Gal. midplane)  we  don't inject small scales, we simply keep the ones observed at high SNR by _Planck_. 

#### Spectra at intermediate and low Gal. latitudes : 
![image](https://user-images.githubusercontent.com/8957029/195357201-2ed3a76f-39cb-4839-83df-ecb55b318857.png)



#### Spectra at high Gal. latitudes 
![image](https://user-images.githubusercontent.com/8957029/195356980-06deb018-66a0-4c3b-b8a1-0f726e83540e.png)

**Comment by brandonshensley** [comment]
Thanks @giuspugl, this is great! One question though: why does the new T map look like it retains less information from the input map relative to the old T map?

**Comment by giuspugl** [comment]
> why does the new T map look like it retains less information from the input map relative to the old T map?

My understanding is that small scale content  in P map is larger especially at high multipoles (because it 's less modulated and because E and B mode spectral indices are flatter) therefore you get a loss of info also in the old T map.

**Comment by zonca** [comment]
@brandonshensley can someone do a further validation with @giuspugl's  maps at `/global/cscratch1/sd/giuspugl/cmbs4/pysm_validation/` before I implement this into PySM?

**Comment by brandonshensley** [comment]
Yes, let's please hold off for now in implementation in PySM. Thanks!

**Comment by seclark** [comment]
Agree on holding off for now. Are there validation measurements needed that don't yet have someone assigned?

**Comment by zonca** [comment]
@giuspugl apologies, I lost the last 10 min of today's call. Are you going to post the notebook here?

**Comment by giuspugl** [comment]
# Updates on dust 
Below a summary of the updates implemented on new dust models : 
- input templates from GNILC variable resolution and unires GNILC maps
- two pivot scales `ell1=110`,`ell2=2000` for small scale injection (see my previous posts ) 
- spectral indices for EE, BB coming from literature , [Planck 2018 XI](https://arxiv.org/pdf/1801.04945.pdf) and for TT from [Miville-Deschenes 2016](https://www.aanda.org/articles/aa/pdf/2016/09/aa28503-16.pdf)
- we inject small scales with non-zero TE from [Planck 2018 XI](https://arxiv.org/pdf/1801.04945.pdf)
- Inside the GAL097 mask (i.e. along the Gal. midplane) we don't inject small scales, we simply keep the ones observed at high SNR by Planck.
- Modulation of qu maps with a single map, p:
 i. to avoid modulation w/ negative values,
 ii. to preserve non-zero TE 
- we propose to modulate small scales as it has been done in `pysm2`(https://arxiv.org/pdf/1608.02841.pdf) ,  with a couple of differences: 
 i. small scales are expected to be injected with   non-gaussian content (thanks to the  logpoltens formalism); 
ii. split the sky with  high reso pixels (nside=8) ; 
iii. use amplitude of E-mode spectra to derive the modulation template

### Validation 
##### TT, TE, EE, BB Power Spectra at different GAL masks 
<img width="749" alt="Screen Shot 2022-11-10 at 5 46 07 PM" src="https://user-images.githubusercontent.com/8957029/201155872-57efa85c-3a7e-4830-b744-4036fc642749.png">


#####  E-to- B ratio at different GAL masks 
![image](https://user-images.githubusercontent.com/8957029/201153801-f8caffa7-156f-40a7-bf04-f15b4eded3d5.png)

#####   EE BB spectra in small patches 
![image](https://user-images.githubusercontent.com/8957029/201154871-d9ea9675-2982-4661-bd81-d9c75ec51d02.png)

##### Polarization Fraction pixel distribution
![image](https://user-images.githubusercontent.com/8957029/201155726-65826521-b184-4665-b89d-3f8e7e913faf.png)

**Comment by zonca** [comment]
thanks @giuspugl! I'll start implementing this into PySM

**Comment by zonca** [comment]
@giuspugl can you please make the data folder `/global/cscratch1/sd/giuspugl/workstation/FG_extension/extending_gnilc_dust/` readable?

**Comment by zonca** [comment]
> @giuspugl can you please make the data folder `/global/cscratch1/sd/giuspugl/workstation/FG_extension/extending_gnilc_dust/` readable?

nevermind, I reran `namaster` on the patches

**Comment by zonca** [comment]
@giuspugl clarification on cell 45:

        ii_LS_alm[ii] = hp.almxfl(alm_log_pol_tens_fullsky[ii], (1.0 - sig_func) ** 0.2)

if we are modulating the small scales in `C_ell` with `sig_func`, shouldn't we modulate these with `sqrt(1 - sig_func)`?

**Comment by giuspugl** [comment]
> @giuspugl clarification on cell 45:
> 
> ```
>     ii_LS_alm[ii] = hp.almxfl(alm_log_pol_tens_fullsky[ii], (1.0 - sig_func) ** 0.2)
> ```
> 
> if we are modulating the small scales in `C_ell` with `sig_func`, shouldn't we modulate these with `sqrt(1 - sig_func)`?

This is something we found in the past to have a _rounder_ transition at around `ell1` in the sigmoid, unfortunately the sqrt was  too sharp.

**Comment by zonca** [comment]
ok, I have modified giuspugl notebook to have a fixed seed and used it to compare with my 2-step implementation (first generate inputs in spherical harmonics space, basically inputs to d11, then from that create the d10 templates), for reference:

https://gist.github.com/c5f71634f4466a52945eb22b0044a3c2
https://gist.github.com/3bf14002bb54d2959c9cb02de48674fe

**Comment by zonca** [comment]
ok, implemented in #133
